ec2/utils.py

```diff
@@ -76,120 +215,267 @@ def get_instance_state(instance, ctx):
-    """ given the related instance object
-        the given variable can be retrieved and returned
-    """
-    timeout = time.time() + timeout_length
+def get_attached_keypair_id(ctx):
+    relationship_type = 'instance_connected_to_keypair'
 
-    while instance.update() != 'running':
-        time.sleep(5)
-        if time.time() > timeout:
-            raise NonRecoverableError('Timed out while attemting to get the '
-                                      'instance {0}. Timeout length: {1}.'
-                                      .format(attribute, timeout_length))
-    attribute = getattr(instance, attribute)
-    return attribute
+    kplist = get_target_aws_resource_id(relationship_type, ctx=ctx)
+    if kplist:
+        return kplist[0]
+    else:
+        return kplist
 
 
-def get_private_dns_name(instance, timeout_length):
-    """ returns the private_dns_name variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'private_dns_name', timeout_length)
+def get_attached_security_group_ids(ctx):
+    relationship_type = 'instance_connected_to_security_group'
 
+    return get_target_aws_resource_id(relationship_type, ctx=ctx)
 
-def get_public_dns_name(instance, timeout_length):
-    """ returns the public_dns_name variable for a given instance
+
+def get_target_aws_resource_id(relationship_type, ctx):
+    """ This loops through the relationships of type and returns
+        targets of those relationships.
     """
-    return get_instance_attribute(instance,
-                                  'public_dns_name', timeout_length)
 
+    ids = []
 
-def get_private_ip_address(instance, timeout_length):
-    """ returns the private_ip_address variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'private_ip_address', timeout_length)
+    if not hasattr(ctx.instance, 'relationships'):
+        ctx.logger.info('Skipping attaching relationships, '
+                        'because none are attached.')
+        return ids
 
+    for r in ctx.instance.relationships:
+        if relationship_type in r.type:
+            ids.append(
+                r.target.instance.runtime_properties['aws_resource_id'])
+
+    return ids
 
-def get_public_ip_address(instance, timeout_length):
-    """ returns the public_ip_address variable for a given instance
-    """
-    return get_instance_attribute(instance,
-                                  'public_ip_address', timeout_length)
 
+def get_security_group_from_id(group_id, ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
+    ctx.logger.debug('Getting Security Group by ID: {0}'.format(group_id))
+
+    try:
+        group = ec2_client.get_all_security_groups(group_ids=group_id)
+    except boto.exception.EC2ResponseError as e:
+        group = get_security_group_from_name(group_id, ctx=ctx)
+        raise NonRecoverableError('Error. '
```

Comment:
yeah i preserved that functionlity using regex which is close to what amazon does